src/math.rs

```diff
@@ -10,6 +12,30 @@ pub(super) fn div_round_up(n: usize, divisor: usize) -> usize {
     }
 }
 
+/// Normalize arbitrary `RangeBounds` to a `Range`
+pub(super) fn simplify_range(range: impl RangeBounds<usize>, len: usize) -> Range<usize> {
+    let start = match range.start_bound() {
+        Bound::Unbounded => 0,
+        Bound::Included(&i) if i <= len => i,
+        Bound::Excluded(&i) if i < len => i + 1,
+        bound => panic!("range start {:?} should be <= length {}", bound, len),
+    };
+    let end = match range.end_bound() {
+        Bound::Unbounded => len,
+        Bound::Excluded(&i) if i <= len => i,
+        Bound::Included(&i) if i < len => i + 1,
```

Comment:
should we worry about overflow? or at least do `check_add(1).unwrap()`?

Reply:
(I feel like if we generally assume no overflow, then we should've made `1..2` and `1..=1` represented by the same data structures...oh well...)

Reply:
With `i < len`, `i + 1` will never overflow.

Reply:
OK, I see, so we *are* doing checked addition, so to speak.

Reply:
Right, we're checking the bounds of the slice, and that implicitly proves the `MAX` bound too.